### PR TITLE
[plugin_platform_interface] Switch `PlatformInterface._instanceToken` to an expando.

### PR DESCRIPTION
**Pull request by stereotype441**
This change replaces `PlatformInterface._instanceToken` (an instance field that points from a `PlatformInterface` to its corresponding token) with an expando that maps from `PlatformInterface` to `Object`. There is no change to the public API, and no change to the behavior of users' production code.

This change ensures that if a customer tries to implement `PlatformInterface` using `implements` rather than `extends`, the code in `PlatformInterface._verify` won't try to access the private `_instanceToken` field in `PlatformInterface`.  This is important because an upcoming change to the dart language is going to cause such accesses to throw exceptions rather than deferring to `noSuchMethod` (see https://github.com/dart-lang/language/issues/2020 for details).

Fixes https://github.com/flutter/flutter/issues/109339.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/main/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/plugin_platform_interface/lib/plugin_platform_interface.dart [hunk 1]
```diff
@@ -44,9 +44,19 @@ abstract class PlatformInterface {
...
   ///
   /// @param token The same, non-`const` `Object` that will be passed to `verify`.
-  PlatformInterface({required Object token}) : _instanceToken = token;
+  PlatformInterface({required Object token}) {
+    _instanceTokens[this] = token;
+  }
 
-  final Object? _instanceToken;
+  /// Expando mapping instances of PlatformInterface to their associated tokens.
+  /// The reason we don't simply use a private field of type `Object?` is
```
Comment (stuartmorgan) [inline packages/plugin_platform_interface/lib/plugin_platform_interface.dart:52]: Nit: could you adjust this comment (here and below) to remove the use of "we"? I can't actually tell who "we" is in this context (see go/avoidwe).

Comment (stereotype441) [reply packages/plugin_platform_interface/lib/plugin_platform_interface.dart]: Done.

File: packages/plugin_platform_interface/lib/plugin_platform_interface.dart [hunk 2]
```diff
@@ -44,9 +44,19 @@ abstract class PlatformInterface {
...
+  /// because as part of implementing
+  /// https://github.com/dart-lang/language/issues/2020, it will soon become a
```
[the remark this answers is not in the document]
Comment (stereotype441) [reply packages/plugin_platform_interface/lib/plugin_platform_interface.dart]: Done.

File: packages/plugin_platform_interface/lib/plugin_platform_interface.dart [hunk 3]
```diff
@@ -44,9 +44,19 @@ abstract class PlatformInterface {
...
+  /// runtime error to invoke a private member that is mocked in another
+  /// library.  By using an expando rather than an instance field, we ensure
+  /// that we will be able to reliably detect implementations using `implements`
```
[the remark this answers is not in the document]
Comment (stereotype441) [reply packages/plugin_platform_interface/lib/plugin_platform_interface.dart]: Done.

**Comment by stereotype441** [comment]
@stuartmorgan, given that this change only affects the behavior of some corner cases in tests, I'm unsure whether I need:
- a CHANGELOG entry
- a version bump
- tests

What do you think?

**Comment by stuartmorgan** [comment]
> @stuartmorgan, given that this change only affects the behavior of some corner cases in tests, I'm unsure whether I need:
> 
> * a CHANGELOG entry
> * a version bump

It's changing production code (and will actually change behavior for clients), so needs both.

> * tests

Since we already have tests that would break later without this change, I think it's covered; I would ping Hixie for an exemption on this.

**Comment by Hixie** [comment]
Do we know of anyone who uses `implements`? Can't we just have them switch to `extends` instead? `Expando` logic is hard to understand and keeping the code simple seems very preferable unless the cost of migrating people is disproportionally large.

**Comment by Hixie** [comment]
Ah, that was not clear from the original commit message or changelog entry. I definitely support the goal here. I just wish there was a cleaner (read, less Expando-magical way) to do it...

File: packages/plugin_platform_interface/lib/plugin_platform_interface.dart [hunk 4]
```diff
@@ -89,10 +100,10 @@ abstract class PlatformInterface {
       return;
     }
     if (preventConstObject &&
-        identical(instance._instanceToken, const Object())) {
```
Comment (Hixie) [inline packages/plugin_platform_interface/lib/plugin_platform_interface.dart:92]: how did this code work before? I don't see a `noSuchMethod` anywhere. Wouldn't this have just thrown at runtime?

Comment (stuartmorgan) [reply packages/plugin_platform_interface/lib/plugin_platform_interface.dart:92]: It was previously wrong (see https://github.com/flutter/flutter/issues/109339 for some more discussion), and the language change was going to make it somewhat differently wrong in at least some cases. This makes it actually do what it was supposed to do, error-wise.

**Comment by Hixie** [comment]
I think this is worth testing, FWIW. Verifying that we give good clear error messages for things like this is valuable.

File: packages/plugin_platform_interface/lib/plugin_platform_interface.dart (hunk 4 above)
Comment (Hixie) [reply packages/plugin_platform_interface/lib/plugin_platform_interface.dart:92]: All the more reason to test it. :-)

Comment (stereotype441) [reply packages/plugin_platform_interface/lib/plugin_platform_interface.dart:92]: Ok, test added!

**Comment by stereotype441** [comment]
> I think this is worth testing, FWIW. Verifying that we give good clear error messages for things like this is valuable.

Test added!

File: packages/plugin_platform_interface/test/plugin_platform_interface_test.dart [hunk 5]
```diff
@@ -21,6 +21,12 @@ class SamplePluginPlatform extends PlatformInterface {
 class ImplementsSamplePluginPlatform extends Mock
     implements SamplePluginPlatform {}
 
+class ImplementsSamplePluginPlatformUsingNoSuchMethod
+    implements SamplePluginPlatform {
+  @override
+  dynamic noSuchMethod(Invocation invocation) => super.noSuchMethod(invocation);
```
Comment (stuartmorgan) [inline packages/plugin_platform_interface/test/plugin_platform_interface_test.dart:27]: Should this implementation maybe be a no-op instead of a call to `super` so we're validating that someone can't completely break the check with a non-standard `noSuchMethod`?

**Review by stuartmorgan — APPROVED** [review]
LGTM; one optional test change.

File: packages/plugin_platform_interface/test/plugin_platform_interface_test.dart (hunk 5 above)
Comment (stereotype441) [reply packages/plugin_platform_interface/test/plugin_platform_interface_test.dart:27]: That might be an interesting test case to add, but it's not the test case that's fixed by this PR.  Even before this change, a `noSuchMethod` that was a no-op (and returned, say `null`) would fail the `if (!identical(token, instance._instanceToken))` check, and cause `_verify` to throw an `AssertionError`, as intended.

Comment (stuartmorgan) [reply packages/plugin_platform_interface/test/plugin_platform_interface_test.dart:27]: Ah, right. Sorry, the exact details here are subtle so they get swapped out pretty quickly for me :)
